src/Microsoft.DocAsCode.MarkdigEngine.Extensions/TripleColon/CodeExtension.cs

```diff
@@ -0,0 +1,258 @@
+            var source = string.Empty;
+            var range = string.Empty;
+            var id = string.Empty;
+            var highlight = string.Empty;
+            var language = string.Empty;
+            var interactive = string.Empty;
+
+            foreach (var attribute in attributes)
+            {
+                var name = attribute.Key;
+                var value = attribute.Value;
+                switch (name)
+                {
+                    case "source":
+                        source = value;
+                        break;
+                    case "range":
+                        range = value;
+                        break;
+                    case "id":
+                        id = value;
+                        break;
+                    case "highlight":
+                        highlight = value;
+                        break;
+                    case "language":
+                        language = value;
+                        break;
+                    case "interactive":
+                        interactive = value;
+                        break;
+                    default:
+                        logError($"Unexpected attribute \"{name}\".");
+                        return false;
+                }
+            }
+
+            if (string.IsNullOrEmpty(source))
+            {
+                logError($"source is a required attribute. Please ensure you have specified a source attribute");
+                return false;
+            }
+            var (code, codePath) = _context.ReadFile(source, InclusionContext.File, null);
+            if(string.IsNullOrEmpty(code))
+            {
+                logError($"The code snippet \"{source}\" could not be found.");
+                return false;
+            }
+            
+            var updatedCode = GetCodeSnippet(range, id, code, logError).TrimEnd('\r').TrimEnd('\n');
+            updatedCodes.Add(updatedCode);
+
+            if (updatedCode == string.Empty)
+            {
+                return false;
+            }
+
+            if(string.IsNullOrEmpty(language))
+            {
+                language = InferLanguageFromFile(source, logError);
+            }
+
+            htmlAttributes = new HtmlAttributes();
+            htmlAttributes.AddProperty("class", $"lang-{language}");
+            if (!string.IsNullOrEmpty(interactive))
+            {
+                htmlAttributes.AddProperty("data-interactive", language);
+                htmlAttributes.AddProperty("data-interactive-mode", interactive);
+            }
+            if (!string.IsNullOrEmpty(highlight)) htmlAttributes.AddProperty("highlight-lines", highlight);
+
+            RenderDelegate = (renderer, obj) =>
+            {
+                renderer.WriteLine("<pre>");
+                renderer.Write("<code").WriteAttributes(obj).Write(">");
+                renderer.WriteLine(updatedCodes.FirstOrDefault());
```

Comment:
In v3, this `CodeExtension` will only have one global instance, and the `ITripleColonExtensionInfo.Render`, `ITripleColonExtensionInfo.TryProcessAttributes` methods will be called on the same instance for many articles in parallel from different threads. The current implementation of `RenderDelegate` captures and mutates the `updatedCodes` variable, this can cause bugs where `updatedCodes` of one article is used to render another article

Reply:
@yufeih - Really good insight, thank you. I made it so things are resolved at the right time and place inside the RenderDelegate.